### ISSUE TITLE
[Week 4] [High] `withdraw_from_yield` applies PnL with inverted sign, yield leaks out of accounting, losses inflate it

### ISSUE THREAD

**Issue by 0xsophon** [Frankcastleauditor/Solana-Audit-Arena#146]
## Finding

**Week**: 4
**Researcher**: 0xsophon (GitHub) + @0xSantii (X)
**Severity**: High
**Category**: Accounting / state corruption - sign error in PnL application
**Affected function**: `withdraw_from_yield::handler`

## Description

After CPI-withdrawing from `yield_generator`, the handler measures realized PnL as the lamport delta on the vault and updates `pool.total_sol` accordingly. The signs are inverted:

```rust
// programs/stake_v2/src/instructions/withdraw_from_yield.rs:76-82
if pnl_i128 >= 0 {
    let gain = u64::try_from(pnl_i128)...?;
    pool.total_sol = PodU64::from(checked_sub_u64(pool.total_sol.get(), gain)?); // -= on PROFIT
} else {
    let loss = u64::try_from(-pnl_i128)...?;
    pool.total_sol = PodU64::from(checked_add_u64(pool.total_sol.get(), loss)?); // += on LOSS
}
```

Correct behavior: a profit should **increase** `total_sol` (because real lamports came back into the pool), a loss should **decrease** it. The current code does the opposite.

## Impact

- **Positive yield (default case):** `pool.total_sol` is decreased instead of increased. The yield SOL sits in the vault but is no longer accounted for. Existing frankSOL holders cannot redeem their share of it — when they unstake they get the price computed from the (artificially low) `total_sol`. New stakers entering after the call get a discount and effectively absorb the yield from old holders.
- **Negative yield (`set_yield_direction(false)`):** `pool.total_sol` is increased instead of decreased. The pool's books overstate solvency: early unstakers cash out at the inflated price and drain the real vault, later unstakers' transactions revert because the vault is empty. Those holders lose access to their stake.

The loss happens every time yield is harvested, accumulates across cycles, and there is no admin instruction to correct `total_sol` after the fact.

## Proof of Concept

Initial pool state (numbers in whole SOL for readability; everything scales to lamports):

```
pool.total_sol      = 1000
pool.deployed_sol   = 0
pool.franksol_supply = 1000     (price = 1.0 SOL per frankSOL)
vault.lamports      = 1000
Bob holds 100 frankSOL pre-existing.
```

Step 1 — fund_manager calls `deploy_to_yield(500)`:
```
vault.lamports     = 500
yield_vault        = 500
pool.deployed_sol  = 500
pool.total_sol     = 1000   (unchanged, money is still ours)
```

Step 2 — ~1 year later, yield_generator credits +50 SOL (10% APY).

Step 3 — fund_manager calls `withdraw_from_yield(principal_returned = 500, yield_amount = 50)`:
- yield_generator transfers 550 lamports → vault.
- `vault_before = 500`, `vault_after = 1050`, `total_return = 550`, `pnl = +50`.
- `pool.deployed_sol -= 500` → 0. 
- **BUG**: `pool.total_sol -= 50` → **950**. (Correct value would be **1050**.)

State after step 3:
```
vault.lamports       = 1050    (real)
pool.total_sol       = 950     (buggy, off by 100 vs reality)
pool.deployed_sol    = 0
pool.franksol_supply = 1000
```

Step 4 — Alice, a new staker, deposits 95 SOL:
- `franksol_out = 95 * 1000 / 950 = 100 frankSOL` (buggy world)
- `franksol_out = 95 * 1000 / 1050 ≈ 90.47 frankSOL` (correct world)
- → Alice mints **+10.5% extra** frankSOL.

Step 5 — Bob unstakes his 100 frankSOL (held since before the yield):
- Buggy state: `total_sol = 1045`, `supply = 1100` → `sol_out = 100 * 1045 / 1100 = 95 SOL`.
- Correct state would have been: `total_sol = 1145`, `supply ≈ 1090.47` → `sol_out ≈ 105 SOL`.
- → Bob receives **95 SOL instead of his fair ~105 SOL — a ~10 SOL loss of yield share.**

After step 5:
```
vault.lamports       = 1050    (still real; only 95 actually paid out)
pool.total_sol       = 950
pool.deployed_sol    = 0
pool.franksol_supply = 1000
```
**100 lamports of yield are stranded** in the vault forever — neither Bob (who left) nor Alice (whose claim against `total_sol` only covers 95) can ever redeem them.

## Recommended Fix

Swap the two arithmetic branches so profit increases and loss decreases `total_sol`. Same `i128` PnL value, just inverted application:

```diff
--- a/programs/stake_v2/src/instructions/withdraw_from_yield.rs
+++ b/programs/stake_v2/src/instructions/withdraw_from_yield.rs
@@ -73,12 +73,12 @@ pub fn handler(
     pool.deployed_sol = PodU64::from(checked_sub_u64(
         pool.deployed_sol.get(),
         principal_returned,
     )?);
     if pnl_i128 >= 0 {
         let gain = u64::try_from(pnl_i128).map_err(|_| StakeError::MathOverflow)?;
-        pool.total_sol = PodU64::from(checked_sub_u64(pool.total_sol.get(), gain)?);
+        pool.total_sol = PodU64::from(checked_add_u64(pool.total_sol.get(), gain)?);
     } else {
         let loss = u64::try_from(-pnl_i128).map_err(|_| StakeError::MathOverflow)?;
-        pool.total_sol = PodU64::from(checked_add_u64(pool.total_sol.get(), loss)?);
+        pool.total_sol = PodU64::from(checked_sub_u64(pool.total_sol.get(), loss)?);
     }
     Ok(())
 }
```

Additional hardening worth considering in the same change:
- Add an invariant assertion `pool.total_sol == vault.lamports + pool.deployed_sol` (modulo rent-exempt baseline) at the end of every state-changing instruction, gated behind a `#[cfg(feature = "audit")]`. This single check would have caught the bug end-to-end.
- The `_yield_amount` parameter passed into the CPI is currently unused for verification on the stake_v2 side. Consider checking `pnl_i128 == yield_amount as i128` (positive case) so a misbehaving yield_generator can't silently lie about the realized return.

**Comment by novoyd** [comment]
This looks like the same inverted-PnL root as #144: the positive and negative yield branches update `pool.total_sol` in the wrong direction. Since #144 was submitted earlier, I’d treat this one as duplicate.

**Comment by Frankcastleauditor** [comment]
**Verdict: ❌ DUPLICATE of #144**
**Score: 0 points**
